### PR TITLE
[TOOLS-2916] Digest Section generator

### PR DESCRIPTION
**Pull request by taksan**
- Digest section generation
- Persisting budget projections in logbook

### CODE REVIEW

File: dashboard-scenario/src/main/java/objective/taskboard/scenario/model/AverageForecastScenarioModel.java [hunk 1]
```diff
@@ -78,7 +79,13 @@ protected WeekParametersModel fetchWeekParameters(long week, LocalDate weekDate)
 
     @Override
     protected LogbookModel scenarioLogbookEntry(LocalDate weekDate) {
-        return new ReadOnlyLogbookModel(project, weekDate, parametersForDate(weekDate), timelineForCurrentWeek());
+        return new ReadOnlyLogbookModel(project, weekDate, parametersForDate(weekDate), timelineForCurrentWeek(), budgetForCurrentWeek());
+    }
+
+    @Override
+    public ScenarioModel withStartSimulationDate(LocalDate weekDate) {
```
Comment (herbert-venancio) [inline dashboard-scenario/src/main/java/objective/taskboard/scenario/model/AverageForecastScenarioModel.java]: Change from `public` to `protected`

File: dashboard-scenario/src/main/java/objective/taskboard/scenario/model/StandardScenarioModel.java [hunk 2]
```diff
@@ -91,9 +92,17 @@ protected WeekParametersModel fetchWeekParameters(long week, LocalDate weekDate)
...
+        BudgetParametersModel budget = budgetForCurrentWeek();
         return parameters.isReadOnly()
-                ? new ReadOnlyLogbookModel(project, weekDate, parameters, timeline)
-                : new EditableLogbookModel(project, weekDate, parameters, timeline);
+                ? new ReadOnlyLogbookModel(project, weekDate, parameters, timeline, budget)
+                : new EditableLogbookModel(project, weekDate, parameters, timeline, budget);
+    }
+
+    @Override
+    public ScenarioModel withStartSimulationDate(LocalDate weekDate) {
```
Comment (herbert-venancio) [inline dashboard-scenario/src/main/java/objective/taskboard/scenario/model/StandardScenarioModel.java]: Change from `public` to `protected`

File: dashboard-scenario/src/test/java/objective/taskboard/scenario/model/ScenarioModelTest.java [hunk 3]
```diff
@@ -242,7 +242,13 @@ public void delete() {}
 
         @Override
         protected LogbookModel scenarioLogbookEntry(LocalDate weekDate) {
-            return new ReadOnlyLogbookModel(project, weekDate, mock(WeekParametersModel.class), mock(TimelineParametersModel.class));
+            return new ReadOnlyLogbookModel(project, weekDate, mock(WeekParametersModel.class), mock(TimelineParametersModel.class), mock(BudgetParametersModel.class));
+        }
+
+        @Override
+        public ScenarioModel withStartSimulationDate(LocalDate weekDate) {
```
Comment (herbert-venancio) [inline dashboard-scenario/src/test/java/objective/taskboard/scenario/model/ScenarioModelTest.java]: Change from `public` to `protected`

File: strategical-dashboard/src/test/java/objective/taskboard/monitor/BudgetSectionGeneratorTest.java [hunk 4]
```diff
@@ -231,11 +232,18 @@ protected LogbookModel scenarioLogbookEntry(LocalDate weekDate) {
             return null;
         }
 
+        @Override
+        public ScenarioModel withStartSimulationDate(LocalDate weekDate) {
```
Comment (herbert-venancio) [inline strategical-dashboard/src/test/java/objective/taskboard/monitor/BudgetSectionGeneratorTest.java]: Change from `public` to `protected`

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 4 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![BLOCKER][BLOCKER] [StrategicalProjectDataSet.java#L58](https://github.com/objective-solutions/taskboard/blob/1da4530464ea446e88124bf91f9700018063e387/dashboard-scenario/src/main/java/objective/taskboard/monitor/StrategicalProjectDataSet.java#L58): Rename field "name" to prevent any misunderstanding/clash with method "name" defined in superclass "java.lang.Enum". [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1845)
1. ![MAJOR][MAJOR] [ChartServiceUtils.java#L14](https://github.com/objective-solutions/taskboard/blob/1da4530464ea446e88124bf91f9700018063e387/dashboard-scenario/src/main/java/objective/taskboard/monitor/ChartServiceUtils.java#L14): Add a private constructor to hide the implicit public one. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1118)
1. ![MAJOR][MAJOR] [BudgetProjectScenarioDSL.java#L336](https://github.com/objective-solutions/taskboard/blob/1da4530464ea446e88124bf91f9700018063e387/dashboard-scenario/src/test/java/objective/taskboard/monitor/BudgetProjectScenarioDSL.java#L336): Swap these 2 arguments so they are in the correct order: expected value, actual value. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS3415)
1. ![MAJOR][MAJOR] [BudgetProjectScenarioDSL.java#L343](https://github.com/objective-solutions/taskboard/blob/1da4530464ea446e88124bf91f9700018063e387/dashboard-scenario/src/test/java/objective/taskboard/monitor/BudgetProjectScenarioDSL.java#L343): Swap these 2 arguments so they are in the correct order: expected value, actual value. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS3415)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: dashboard-scenario/src/main/resources/db/migration/dashboard/scenario/oracle/V201909251054__createBudgetParametersTable.sql [hunk 5]
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE budget_parameters (
+    id              NUMBER    PRIMARY KEY
+  , created         TIMESTAMP NOT NULL
+  , updated         TIMESTAMP NOT NULL
+  , budget_conservative DECIMAL(5,4) NULL
```
Comment (herbert-venancio) [inline dashboard-scenario/src/main/resources/db/migration/dashboard/scenario/oracle/V201909251054__createBudgetParametersTable.sql]: Change from `budget_conservative` to `conservative_budget`

File: dashboard-scenario/src/main/resources/db/migration/dashboard/scenario/oracle/V201909251054__createBudgetParametersTable.sql [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+  , budget_optimistic   DECIMAL(5,4) NULL
```
Comment (herbert-venancio) [inline dashboard-scenario/src/main/resources/db/migration/dashboard/scenario/oracle/V201909251054__createBudgetParametersTable.sql]: Change from `budget_optimistic` to `optimistic_budget`

File: dashboard-scenario/src/main/resources/db/migration/dashboard/scenario/oracle/V201909251054__createBudgetParametersTable.sql (hunk 5 above)
Comment (herbert-venancio) [inline dashboard-scenario/src/main/resources/db/migration/dashboard/scenario/oracle/V201909251054__createBudgetParametersTable.sql]: Change from `DECIMAL(5,4)` to `DECIMAL(9,3)`

File: dashboard-scenario/src/main/resources/db/migration/dashboard/scenario/oracle/V201909251054__createBudgetParametersTable.sql (hunk 6 above)
Comment (herbert-venancio) [inline dashboard-scenario/src/main/resources/db/migration/dashboard/scenario/oracle/V201909251054__createBudgetParametersTable.sql]: Change from `DECIMAL(5,4)` to `DECIMAL(9,3)`

File: dashboard-scenario/src/main/resources/db/migration/dashboard/scenario/mysql/V201909251008__createBudgetParametersTable.sql [hunk 7]
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE `budget_parameters` (
+    `id` bigint(20) not null auto_increment,
+    `created`         DATETIME(3) NOT NULL,
+    `updated`         DATETIME(3) NOT NULL,
+    `conservative_budget` bigint(20) not null,
```
Comment (herbert-venancio) [inline dashboard-scenario/src/main/resources/db/migration/dashboard/scenario/mysql/V201909251008__createBudgetParametersTable.sql]: Change from `bigint(20)` to `DECIMAL(9,3)`

File: dashboard-scenario/src/main/resources/db/migration/dashboard/scenario/mysql/V201909251008__createBudgetParametersTable.sql [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+    `optimistic_budget` bigint(20) not null,
```
Comment (herbert-venancio) [inline dashboard-scenario/src/main/resources/db/migration/dashboard/scenario/mysql/V201909251008__createBudgetParametersTable.sql]: Change from `bigint(20)` to `DECIMAL(9,3)`

File: dashboard-scenario/src/test/java/objective/taskboard/scenario/model/AverageForecastScenarioModelTest.java [hunk 9]
```diff
@@ -254,7 +254,7 @@ void initMocks() {
                 projectScenario.getLogbooks().add(logbook);
                 logbook.setId(idGenerator.incrementAndGet());
                 return logbook;
-            }).given(scenarioService).createLogbook(any(), any(), any(), any());
+            }).given(scenarioService).createLogbook(any(), any(), any(), any(), any());
```
Comment (herbert-venancio) [inline dashboard-scenario/src/test/java/objective/taskboard/scenario/model/AverageForecastScenarioModelTest.java]:
```Java
            willAnswer(invocation -> {
                ProjectScenario projectScenario = invocation.getArgumentAt(0, ProjectScenario.class);
                LocalDate weekDate = invocation.getArgumentAt(1, LocalDate.class);
                WeekParameters weekParameters = invocation.getArgumentAt(2, WeekParameters.class);
                TimelineParameters timelineParameters = invocation.getArgumentAt(3, TimelineParameters.class);
                BudgetParameters budgetParameters = invocation.getArgumentAt(4, BudgetParameters.class); // <-- add this

                weekParameters.setId(idGenerator.incrementAndGet());
                timelineParameters.setId(idGenerator.incrementAndGet());
                Logbook logbook = new Logbook();
                logbook.setWeek(weekDate);
                logbook.setWeekParameters(weekParameters);
                logbook.setTimelineParameters(timelineParameters);
                logbook.setBudgetParameters(budgetParameters); // <-- add this
                logbook.setProjectScenario(projectScenario);
                projectScenario.getLogbooks().add(logbook);
                logbook.setId(idGenerator.incrementAndGet());
                return logbook;
            }).given(scenarioService).createLogbook(any(), any(), any(), any(), any());
```

File: strategical-dashboard/src/test/java/objective/taskboard/monitor/util/SectionAssert.java [hunk 10]
```diff
@@ -43,8 +46,13 @@ public SectionAssert numberOfRowsIs(int size) {
     }
 
     public SectionAssert row(int index, String[] expectedValues) {
-        assertThat(actual.bodyRows.get(index))
+        List<List<String>> stringContent = actual.bodyRows.stream()
+                .map(itens -> itens.stream().map(item -> item.content).collect(Collectors.toList()))
```
Comment (herbert-venancio) [inline strategical-dashboard/src/test/java/objective/taskboard/monitor/util/SectionAssert.java]: Change from ite***n***s to ite***m***s

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/DeltaSectionGenerator.java [hunk 11]
```diff
@@ -56,7 +56,7 @@ public SectionBuilder(LogbookModel logbook) {
         }
 
         public LogbookSectionTableDto build() {
-            return new LogbookSectionTableDto(
+            return new LogbookSectionTableDto(2,
```
Comment (herbert-venancio) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/DeltaSectionGenerator.java]: `priority` should be parametrized
Also, consider moving priority to the factory, so it only needs to order once

**Pull request merged**